fpsd/features.py

```diff
@@ -0,0 +1,765 @@
+        """
+
+        num_ranks = num_features + 1
+
+        self.drop_stale_feature_table("features.cell_ordering_differences")
+
+        self._create_temp_packet_positions(outgoing_only=True)
+        self._create_table_outgoing_cell_positions(num_cells=num_ranks)
+
+        self.drop_stale_feature_table("top_{}_cell_ordering".format(num_ranks))
+
+        crosstab_columns = ['outgoing_cell_position_{} bigint'.format(x)
+                            for x in range(1, num_ranks + 1)]
+
+        query = ("CREATE TEMP TABLE top_{n}_cell_ordering AS         "
+                 "(SELECT * FROM crosstab(                           "
+                 "'SELECT * FROM top_{n}_outgoing_packet_positions') "
+                 "AS ct(exampleid integer,                           "
+                 "{cols})); ").format(n=num_ranks,
+                                      cols=', '.join(crosstab_columns))
+        self.execute_query_from_string(query)
+
+        diff_columns = [("({prefix}_{n2} - {prefix}_{n1}) "
+                         "AS {prefix}_difference_{n1}").format(
+                            prefix='outgoing_cell_position', n2=x+1, n1=x)
+                        for x in range(1, num_ranks)]
+
+        feat_columns = ', '.join(diff_columns)
+        query = ("CREATE TABLE features.cell_ordering_differences   "
+                 "AS (SELECT exampleid, {cols}                      "
+                 "FROM top_{n}_cell_ordering); ".format(cols=feat_columns,
+                                                        n=num_ranks))
+
+        self.execute_query_from_string(query)
+        return "features.cell_ordering_differences"
+
+    def generate_table_binned_counts(self, num_features=100, size_window=30):
+        """This method takes all examples and produces a
+        table with the following format:
+
+        exampleid  | a         | b         ...    | num_features
+        (integer)  | (bigint)  | (bigint)  ...    | (bigint)
+        --------------------------------------------------------
+        9          | 11        | 11        ...
+        10         | 12        | 14        ...
+        11         | 13        | 16        ...
+
+        where there are a variable number (num_features) of
+        feature columns, each named num_outgoing_packets_in_window_x_of_size_y.
+        num_outgoing_packets_in_window_x_of_size_y is the number
+        of cells in the xth bin or "window" of size y
+
+        Args:
+            num_features [int]: number of features to create
+            size_window [int]: size of each bin
+
+        Returns:
+            [string] name of newly created table
+        """
+
+        self._create_temp_packet_positions(outgoing_only=False)
+        feature_table_name = "features.size_{}_windows".format(size_window)
+
+        self.drop_stale_feature_table(feature_table_name)
+
+        if num_features > 1:
+            feature_columns = ["num_outgoing_packets_in_window_{}_of_size_{}".format(x,
```

Comment:
Ah, so I’m using dynamic names in cases like this so that if one wants to generate multiple feature tables with a different e.g. number of windows or window sizes and use both of them, then it’s possible to do so.